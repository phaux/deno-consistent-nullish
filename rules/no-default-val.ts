import { trimRange } from "../util/trimRange.ts";

export const noDefaultValRule: Deno.lint.Rule = {
  create: (ctx) => ({
    AssignmentPattern(node) {
      ctx.report({
        range: trimRange(ctx.sourceCode, [
          node.left.range[1],
          node.right.range[0],
        ]),
        message: "Default values are not allowed.",
        hint: "Default value is only applied when the original is undefined but not null. " +
          "Use nullish coalescing operator (`??`) to apply default value.",
      });
    },
  }),
};
