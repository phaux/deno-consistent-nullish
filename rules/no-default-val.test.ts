import { assertSnapshot } from "@std/testing/snapshot";
import { noDefaultValRule } from "./no-default-val.ts";
import { assertEquals } from "@std/assert";

const plugin: Deno.lint.Plugin = {
  name: "test",
  rules: { rule: noDefaultValRule },
};

Deno.test("bans default parameters", async (t) => {
  const diagnostics = Deno.lint.runPlugin(
    plugin,
    "test.ts",
    `
    function foo(a = 1) {}
    const bar = (a = 1, b = 2) => {};
    const baz = function (a = "a") {};
    const obj = { foo(a = {}) {} };
    class Foo { foo(a = []) {} }
    `
  );
  assertEquals(diagnostics.length, 6);
  await assertSnapshot(t, diagnostics);
});

Deno.test("bans default object properties", async (t) => {
  const diagnostics = Deno.lint.runPlugin(
    plugin,
    "test.ts",
    `
    function foo({ a = 1 }) {}
    const bar = ({ a = 1, b = 2 }) => {};
    const baz = function ({ a: a = "a" }) {};
    const { a: b = {} } = {};
    `
  );
  assertEquals(diagnostics.length, 2); // BUG: Deno parses `{ a = b }` as `{ a: b }`
  await assertSnapshot(t, diagnostics);
});

Deno.test("allows object property renames", () => {
  const diagnostics = Deno.lint.runPlugin(
    plugin,
    "test.ts",
    `
    function foo({ a: a }) {}
    const bar = ({ a: b, b: a }) => {};
    const baz = function ({ a: x }) {};
    const { a: x } = {};
    `
  );
  assertEquals(diagnostics.length, 0);
});

Deno.test("bans default array elements", async (t) => {
  const diagnostics = Deno.lint.runPlugin(
    plugin,
    "test.ts",
    `
    function foo([a = 1]) {}
    const bar = ([a = 1, b = 2]) => {};
    const baz = function ([a = "a"]) {};
    const [a = []] = [];
    `
  );
  assertEquals(diagnostics.length, 5);
  await assertSnapshot(t, diagnostics);
});
