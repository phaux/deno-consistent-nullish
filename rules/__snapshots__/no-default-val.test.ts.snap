export const snapshot = {};

snapshot[`bans default parameters 1`] = `
[
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      20,
      21,
    ],
  },
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      47,
      48,
    ],
  },
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      54,
      55,
    ],
  },
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      94,
      95,
    ],
  },
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      129,
      130,
    ],
  },
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      163,
      164,
    ],
  },
]
`;

snapshot[`bans default object properties 1`] = `
[
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      107,
      108,
    ],
  },
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      137,
      138,
    ],
  },
]
`;

snapshot[`bans default array elements 1`] = `
[
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      21,
      22,
    ],
  },
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      50,
      51,
    ],
  },
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      57,
      58,
    ],
  },
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      99,
      100,
    ],
  },
  {
    fix: [],
    hint: "Default value is only applied when the original is undefined but not null. Use nullish coalescing operator (\`??\`) to apply default value.",
    id: "test/rule",
    message: "Default values are not allowed.",
    range: [
      124,
      125,
    ],
  },
]
`;
